### PR TITLE
Fix and rework gitlab_project_variable

### PR DESCRIPTION
**Pull request by markuman**
##### SUMMARY
Closes: #1952 
Closes: #4074

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
gitlab_project_variable

##### ADDITIONAL INFORMATION

First tests looks promising.

```yml
    - name: previous way. still works
      community.general.gitlab_project_variable:
        api_url: https://gitlab.com
        api_token: "{{ some_token }}"
        project: markuman/dotfiles
        purge: false
        vars:
          ACCESS_KEY_ID: abc123
          SECRET_ACCESS_KEY:
            value: 3214cbad
            masked: true
            protected: true
            variable_type: env_var
            environment_scope: '*'

    - name: new way, works also
      community.general.gitlab_project_variable:
        api_url: https://gitlab.com
        api_token: "{{ some_token }}"
        project: markuman/dotfiles
        purge: false
        variables:
          - name: ACCESS_KEY_ID
            value: abc123
          - name: SECRET_ACCESS_KEY
            value: 3214cbad
            masked: true
            protected: true
            variable_type: env_var
            environment_scope: '*'
```

What's left is

- [x] append integration test
- [x] `return_value[untouched]` is not handled
- [x] changelog fragment

### CODE REVIEW

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 1]
```diff
@@ -162,104 +229,91 @@ def list_all_project_variables(self):
...
+    existing_variables = [x.attributes for x in gitlab_keys]
+
+    # preprocessing:filter out and enrich before compare
+    for item in existing_variables:
+        item.pop('project_id')
+
+    for item in requested_variables:
+        item['key'] = item.pop('name')
+        item['value'] = str(item.get('value'))
+        if item.get('protected') is None:
```
Comment (markuman) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:271]: this does not break backwards compatibility.  
but it might result in unexpected behaviours!

Even if we do not enrich the properties with default values, it will also result in unexpected behaviours for the users. Because when they are missing, it is always detect as a change.

I have no idea how we can save the situation...  

cc @felixfontein

Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:272]: Ok, I've changed it a little so that it fits to the current behaviour.  

All what's left to make integration test pass, is the identification of `untouched` variables.

**Comment by markuman** [comment]
Integration test pass for me

```
PLAY RECAP *********************************************************************
testhost                   : ok=70   changed=27   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0  
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 2]
```diff
@@ -55,6 +55,42 @@
         See GitLab documentation on acceptable values for a masked variable (https://docs.gitlab.com/ce/ci/variables/#masked-variables).
     default: {}
     type: dict
+  variables:
+    description:
+      - List of dicts
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:60]:
```suggestion
      - List of dicts.
    version_added: 4.4.0
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 3]
```diff
@@ -55,6 +55,42 @@
...
+        type: bool
+        default: false
+      protected:
+        description:
+          - Wether variable value is protected or not.
+        type: bool
+        default: false
+      variable_type:
+        description:
+          - Wether a variable is an env or a file.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:85]:
```suggestion
          - Wether a variable is an environment variable (C(env_var)) or a file (C(file)).
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Some first comments:

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 4]
```diff
@@ -303,9 +397,29 @@ def main():
...
+        item.pop('project_id')
+        item['name'] = item.pop('key')
+    for item in before:
+        item.pop('project_id')
+        item['name'] = item.pop('key')
+
+    diff = dict(
+        before=yaml.safe_dump(before),
+        after=yaml.safe_dump(after)
+    )
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:413]: Pre-formatting diffs is not a good idea. It's better to let the callback plugin do this. Also there's no reason why users should have PyYAML installed, and this code will crash if it isn't installed.
```suggestion
    diff = dict(
        before=before,
        after=after,
    )
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 5]
```diff
@@ -55,6 +55,43 @@
         See GitLab documentation on acceptable values for a masked variable (https://docs.gitlab.com/ce/ci/variables/#masked-variables).
     default: {}
     type: dict
+  variables:
+    description:
+      - List of dicts.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:60]: That's not a helpful description about what this actually is :-)

Also, it should point out the differences to `vars`, and the `vars` documentation should mention `variables` as well.

File: changelogs/fragments/4038-fix-and-rework-gitlb-project-variable.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+bugfixes:
+  - gitlab_project_variable - allow to set same variable name under different environment scopes (https://github.com/ansible-collections/community.general/issues/4038).
```
Comment (felixfontein) [inline changelogs/fragments/4038-fix-and-rework-gitlb-project-variable.yml:2]: While this new feature contains bugfixes, it is primarily a new feature. It needs a `minor_changes` entry as well (mentioning the new option).

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 7]
```diff
@@ -142,6 +190,39 @@
 from ansible_collections.community.general.plugins.module_utils.gitlab import auth_argument_spec, gitlab_authentication
 
 
+def vars_to_variables(vars):
+    # transform old vars to new variables structure
+    variables = list()
+    for item in vars.keys():
+        if (isinstance(vars.get(item), str) or
+           isinstance(vars.get(item), int) or
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:198]:
```suggestion
        if (isinstance(vars.get(item), string_types) or
           isinstance(vars.get(item), integer_types) or
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 8]
```diff
@@ -142,6 +190,39 @@
...
+           isinstance(vars.get(item), float)):
+            variables.append(
+                {
+                    "name": item,
+                    "value": str(vars.get(item))
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:203]: The old code did not convert values to string. This probably causes a change in behavior of the module.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 9]
```diff
@@ -142,6 +190,39 @@
...
+                }
+            )
+        else:
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:206]: The old code checked whether it was a dictionary, or called `module.fail_json` otherwise. The new code should do the same.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 10]
```diff
@@ -162,106 +243,111 @@ def list_all_project_variables(self):
...
+            for item in existing_variables:
+                if this_gitlab.delete_variable(item):
+                    return_value['removed'].append(item)
+
+    if len(return_value['added'] + return_value['removed'] + return_value['updated']) > 0:
+        change = True
+
+    gitlab_keys = this_gitlab.list_all_project_variables()
+    after = [x.attributes for x in gitlab_keys]
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:348]: I don't think this works in check mode.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution!

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 8 above)
Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:203]: Ah, then this PR fixes another bug

```yml
    - name: same vars, diff scope
      community.general.gitlab_project_variable:
        api_url: https://gitlab.com
        project: markuman/dotfiles
        purge: yes
        vars:
          int_value: 12
      register: bug

    - debug:
        var: bug is changed
       
    - name: same vars, diff scope
      community.general.gitlab_project_variable:
        api_url: https://gitlab.com
        project: markuman/dotfiles
        purge: yes
        vars:
          int_value: 12
    
    - debug:
        var: bug is changed
```

results in

```
TASK [Gathering Facts] ********************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [same vars, diff scope] **************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [debug] ******************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "bug is changed": true
}

TASK [same vars, diff scope] **************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [debug] ******************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "bug is changed": true
}

PLAY RECAP ********************************************************************************************************************************************************************************************************************
localhost                  : ok=5    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

GitLab returns all ENV variables always as string. Even if you put an int or float value to the API.  

The PR version results in 

```

TASK [debug] ******************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "bug is changed": false
}

TASK [same vars, diff scope] **************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [debug] ******************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "bug is changed": false
}

PLAY RECAP ********************************************************************************************************************************************************************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

First task is also not changed because the value exists from the previous run.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 10 above)
Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:348]: Do mean because `before` and `after` will have the same values in check mode?

That do you suggest instead? No return of diff in check mode?

**Comment by markuman** [comment]
Integration test still pass here

```
testhost                   : ok=72   changed=26   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 11]
```diff
@@ -55,6 +55,44 @@
         See GitLab documentation on acceptable values for a masked variable (https://docs.gitlab.com/ce/ci/variables/#masked-variables).
     default: {}
     type: dict
+  variables:
```
Comment (markuman) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:58]: maybe we should mark the `vars` parameter as deprecated and remove it in 6.0.0? @felixfontein

Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:58]: I'd only deprecate it after the new parameter has been around for some time. This gives users a bit of time to migrate without having to see deprecation warnings all the time (think of playbooks that have to run with different versions of the collections).

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 10 above)
Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:345]: > Do mean because `before` and `after` will have the same values in check mode?

Yes. In the old version this was correctly returned also in check mode (IIRC).

> That do you suggest instead? No return of diff in check mode?

I would probably compute the value the same way as it was computed in the previous version.

Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:345]: > In the old version ...

The old version was spaghetti code from hell :D But I see what I can do.

Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:345]: The behaviour is worth discussing 

Say there are two variables: same name, but different scrope.  
Now we request a third variable in purge mode: also the same name, and also a different env scope.

The current implementation counts one delete and one update

```
ok: [localhost] => {
    "gitlab_project_variable_state": {
        "changed": true,
        "failed": false,
        "project_variable": {
            "added": [],
            "removed": [
                "SECRET_ACCESS_KEY"
            ],
            "untouched": [],
            "updated": [
                "SECRET_ACCESS_KEY"
            ]
        }
    }
}
```

Technical that's impossible. We can change the value, but nothing else.  
The new implementation I got now, counts two removals and one add - what is technical correct. The current implementation is a bug.


But I think we want to keep the bug behaviour of an update state for the users, right? :D

Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:345]: It's not possible to implement the `updated` count like it was before. Because the current implementation does not respect different environment scopes using the same variable name. Therefore the `update` might happen in the future. But in the example above, the current implementation is a bug.

Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:345]: The changed behavior sounds good to me. It should be documented in the changelog though.

**Comment by markuman** [comment]
@felixfontein do we need to pop the `value` key from the `before` and `after` return values? Or are threre some other `no_log` techniques available?

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 12]
```diff
@@ -271,7 +354,8 @@ def main():
         project=dict(type='str', required=True),
         purge=dict(type='bool', required=False, default=False),
         vars=dict(type='dict', required=False, default=dict(), no_log=True),
-        state=dict(type='str', default="present", choices=["absent", "present"])
+        variables=dict(type='list', elements='dict', required=False, default=list(), no_log=True),
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:392]: BTW, you need to fill the suboptions in (and use `no_log=True` at the approriate place - making the whole list `no_log` is not appropriate):
```suggestion
        variables=dict(type='list', elements='dict', required=False, default=list(), options=dict(
            ...
        )),
```

**Comment by felixfontein** [comment]
@markuman ansible-core only censores values that show up in parameters marked with `no_log=True`. So if you have secrets that can be retrieved by the API in variables that aren't added/set by this module's invocation, they will not be censored. Maybe removing the values would be best, but on the other hand this makes the diff feature rather useless.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 13]
```diff
@@ -55,6 +55,44 @@
         See GitLab documentation on acceptable values for a masked variable (https://docs.gitlab.com/ce/ci/variables/#masked-variables).
     default: {}
     type: dict
+  variables:
+    version_added: 4.4.0
+    description:
+      - A list of dictionries that represents CI/CD variables.
+      - This modules works internal with this sructure, even if the older I(vars) parameter is used.
```
Comment (sguarin) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:62]: dictionaries
and
structure

**Comment by markuman** [comment]
Integration test pass for me

```
testhost                   : ok=72   changed=27   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 14]
```diff
@@ -142,6 +185,42 @@
 from ansible_collections.community.general.plugins.module_utils.gitlab import auth_argument_spec, gitlab_authentication
 
 
+def vars_to_variables(vars, module):
+    # transform old vars to new variables structure
+    variables = list()
+    for item in vars.keys():
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:191]:
```suggestion
    for item, value in vars.items():
```
and replace every `vars.get(item)` below by `value`.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 15]
```diff
@@ -162,106 +241,147 @@ def list_all_project_variables(self):
...
         if self._module.check_mode:
-            return
+            return True
         var = {
-            "key": key, "value": value,
-            "masked": masked, "protected": protected,
-            "variable_type": variable_type,
+            "key": var_obj.get('key'), "value": var_obj.get('value'),
+            "masked": var_obj.get('masked'), "protected": var_obj.get('protected'),
+            "variable_type": var_obj.get('variable_type'), "environment_scope": var_obj.get('environment_scope')
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:250]: IIRC some versions barf if `environment_scope` is passed in. So better set it only when needed, like the code did before.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 16]
```diff
@@ -162,106 +241,147 @@ def list_all_project_variables(self):
...
+    untouched = list()
+    updated = list()
+    added = list()
+
+    if state == 'present':
+        existing_key_scope_vars = list()
+        for item in existing_variables:
+            existing_key_scope_vars.append({'key': item.get('key'), 'environment_scope': item.get('environment_scope')})
+
+        for idx in range(len(requested_variables)):
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:286]:
```suggestion
        for var in requested_variables:
```
and replace `requested_variables[idx]` by `var` below.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 17]
```diff
@@ -162,106 +241,147 @@ def list_all_project_variables(self):
...
-                change = True
-                return_value['removed'].append(key)
+    if module.check_mode:
+        return_value = dict(added=added, updated=updated, removed=return_value['removed'], untouched=untouched)
 
-        elif key not in existing_variables and state == 'present':
-            this_gitlab.create_variable(key, value, masked, protected, variable_type, environment_scope)
-            change = True
-            return_value['added'].append(key)
+    if len(return_value['added'] + return_value['removed'] + return_value['updated']) > 0:
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:378]: More Pythonic:
```suggestion
    if return_value['added'] or return_value['removed'] or return_value['updated']:
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 18]
```diff
@@ -162,106 +241,147 @@ def list_all_project_variables(self):
...
-    existing_variables = list(filter(None, existing_variables))
-    if purge:
-        for item in existing_variables:
-            this_gitlab.delete_variable(item)
-            change = True
-            return_value['removed'].append(item)
-    else:
-        return_value['untouched'].extend(existing_variables)
+    gitlab_keys = this_gitlab.list_all_project_variables()
+    after = [x.attributes for x in gitlab_keys]
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:382]: Diff stlil does not work in check mode, right?

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 19]
```diff
@@ -162,106 +241,147 @@ def list_all_project_variables(self):
...
+    # preprocessing:filter out and enrich before compare
+    for item in existing_variables:
+        item.pop('project_id')
+
+    for item in requested_variables:
+        item['key'] = item.pop('name')
+        item['value'] = str(item.get('value'))
+        if item.get('protected') is None:
+            item['protected'] = False
+            module.deprecate("The 'protected' attribute will be set to 'True' if not defined.",
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:319]:
```suggestion
            module.deprecate("The 'protected' attribute will be set to 'true' if not defined.",
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 20]
```diff
@@ -162,106 +241,147 @@ def list_all_project_variables(self):
...
+                             version='5.0.0', collection_name='community.general')
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:320]:
```suggestion
                             version='6.0.0', collection_name='community.general')
```
The changelog fragment needs a `deprecated_features` part then.

I don't really get why protected should be set to `true` if not defined? What is the rationale behind that?

Also, why does this have to be done in this PR, which already changes quite a lot? Is this really necessary?

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 21]
```diff
@@ -282,6 +410,7 @@ def main():
             ['api_username', 'api_job_token'],
             ['api_token', 'api_oauth_token'],
             ['api_token', 'api_job_token'],
+            ['vars', 'variables']
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:413]:
```suggestion
            ['vars', 'variables'],
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 22]
```diff
@@ -142,6 +185,42 @@
...
+
+            if vars.get(item).get('variable_type'):
+                new_item['variable_type'] = vars.get(item).get('variable_type')
+
+            variables.append(new_item)
+
+        else:
+            module.fail_json(msg="value must be of type string, integer, float or dict")
+
+    return variables
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:221]: Why not return full dictionaries (with all keys), like the new option does?

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 23]
```diff
@@ -162,106 +241,147 @@ def list_all_project_variables(self):
...
+    # updated => name and scope are equal
+    # added => name and scope does not exist
+    untouched = list()
+    updated = list()
+    added = list()
+
+    if state == 'present':
+        existing_key_scope_vars = list()
+        for item in existing_variables:
+            existing_key_scope_vars.append({'key': item.get('key'), 'environment_scope': item.get('environment_scope')})
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:284]:
```suggestion
            existing_key_scope_vars.append({'key': item.get('key'), 'environment_scope': item.get('environment_scope', '*')})
```
Otherwise you will get idempotency problems if you have a version of the library / GH instance that does not support environment scopes.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 20 above)
Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:320]: > I don't really get why protected should be set to true if not defined? What is the rationale behind that?

It's the Web GUI default behaviour and kind of a security feature (_mitigation of accidental credentials leak on unprotected branches, such as merge requests_)

> Also, why does this have to be done in this PR, which already changes quite a lot? Is this really necessary?

Sure, we can change this in a later PR.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 23 above)
Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:284]: `'*'` is the default scope if you do not defined one. Without using it, you get wrong `changed` return values, because `None != '*'` will always return a difference.

>  if you have a version of the library / GH instance that does not support environment scopes.

That's a good point.  
But it should have been checked at the time where `environment_scope` and `env_type` where introduced. I think we missed it in the past, because only the attributes `protected` and `masked` are documentated.

> Support for protected values requires GitLab >= 9.3.  
> Support for masked values requires GitLab >= 11.10.

About `environment_scope`

> [Support for environment scopes introduced in GitLab Premium 13.11 ](https://docs.gitlab.com/ee/ci/variables/#add-a-cicd-variable-to-a-group)

About `variable_type`

> [Introduced in GitLab 11.11. ](https://docs.gitlab.com/ee/ci/variables/#cicd-variable-types)


Basically that are two more bugs
* lack of documentation, because it is already introduced
* environment_scope is not usable on GitLab CE

I'm not sure if we should target also these two bugs in this PR?

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 22 above)
Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:221]: Do you mean the first `if` block

```py
if (isinstance(vars.get(item), string_types) or
           isinstance(vars.get(item), (integer_types, float))):
```

and why it don't enricht the default properties/attributes?

Good question, I must think about why I did it that way.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 18 above)
Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:382]: Ah yes, I must think about it

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 15 above)
Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:250]: When it is not set, it's `'*'` by GitLab.  
But yes, because of https://github.com/ansible-collections/community.general/pull/4038#discussion_r793349327 it might result in exception on GitLab CE.  
But I'm not sure if a GitLab CE instance quietly accept it, but you just won't see it in the GUI and it has no effect on your pipelines.

I will try against a GitLab CE version later.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 23 above)
Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:284]: I *think* the code was written so that if you don't use `environment_scope` and `variable_type`, it will just work. So it's not really a bug, just a documentation issue.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 15 above)
Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:250]: Some folks might also run very old versions of GitLab (for whatever reason) that were working fine with the code before this PR. We can require a newer version at some point, but not in a minor release.

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 18 above)
Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:382]: The combination of diff and check mode is usually the hardest to implement, but the most useful for users ;-)

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 23 above)
Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:284]: Ok.  
But in that case, for compare, we must work with the default value if `environment_scope` is not requestes. Otherwise it results always a changed task.

**Comment by markuman** [comment]
All integration tests still pass here

```
testhost                   : ok=70   changed=27   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 24]
```diff
@@ -48,13 +48,55 @@
...
+        type: str
+        required: true
+      value:
+        description:
+          - The variable value.
+        type: str
+        required: true
+      masked:
+        description:
+          - Wether variable value is masked or not.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:81]:
```suggestion
          - Whether the variable value is masked or not.
          - Support for masked values requires GitLab >= 11.10.
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 25]
```diff
@@ -48,13 +48,55 @@
...
+        type: bool
+        default: false
+      protected:
+        description:
+          - Wether variable value is protected or not.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:86]:
```suggestion
          - Whether the variable value is protected or not.
          - Support for protected values requires GitLab >= 9.3.
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 26]
```diff
@@ -48,13 +48,55 @@
...
+        default: false
+      variable_type:
+        description:
+          - Wether a variable is an environment variable (C(env_var)) or a file (C(file)).
+        type: str
+        choices: ["env_var", "file"]
+        default: env_var
+      environment_scope:
+        description:
+          - The scope for the variable.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:97]:
```suggestion
          - The scope for the variable.
          - Support for I(environment_scope) requires GitLab Premium >= 13.11.
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 27]
```diff
@@ -48,13 +48,55 @@
...
+        type: bool
+        default: false
+      protected:
+        description:
+          - Wether variable value is protected or not.
+        type: bool
+        default: false
+      variable_type:
+        description:
+          - Wether a variable is an environment variable (C(env_var)) or a file (C(file)).
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:91]:
```suggestion
          - Wether a variable is an environment variable (C(env_var)) or a file (C(file)).
          - Support for I(variable_type) requires GitLab >= 11.11.
```

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 28]
```diff
@@ -142,6 +187,45 @@
...
+            if value.get('protected'):
+                new_item['protected'] = value.get('protected')
+
+            if value.get('environment_scope'):
+                new_item['environment_scope'] = value.get('environment_scope')
+
+            if value.get('variable_type'):
+                new_item['variable_type'] = value.get('variable_type')
+
+            variables.append(new_item)
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:221]: Shouldn't this be
```suggestion
            new_item = {
                "name": item,
                "value": value.get('value'),
                "masked": value.get('masked'),
                "protected": value.get('protected'),
                "variable_type": value.get('variable_type'),
            }

            if value.get('environment_scope'):
                new_item['environment_scope'] = value.get('environment_scope')

            variables.append(new_item)
```
so it behaves similar to the above case?

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py [hunk 29]
```diff
@@ -162,106 +246,150 @@ def list_all_project_variables(self):
...
+    if state == 'present':
+        existing_key_scope_vars = list()
+        for item in existing_variables:
+            existing_key_scope_vars.append({'key': item.get('key'), 'environment_scope': item.get('environment_scope')})
+
+        for var in requested_variables:
+            if var in existing_variables:
+                untouched.append(var)
+            else:
+                compare_item = {'key': var.get('name'), 'environment_scope': var.get('environment_scope', '*')}
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project_variable.py:300]: If in the above code the default value `'*'` must not be there, why must it be here?
```suggestion
                compare_item = {'key': var.get('name'), 'environment_scope': var.get('environment_scope')}
```

Comment (markuman) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py:300]: Ah yes. I think I implement it like this because of [this](https://github.com/ansible-collections/community.general/pull/4038#discussion_r793308583).

I re-run the integration test and see if all sill pass.

**Comment by markuman** [comment]
Still no failures

```
testhost                   : ok=70   changed=26   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

FYI, the `changed` count flipps between 26 and 27 because of the first task, which purges all variables from the test project. If it is already empty, it's one change less

**Comment by markuman** [comment]
hm codespace put the changes somewhere else ...

File: plugins/modules/source_control/gitlab/gitlab_project_variable.py (hunk 29 above)
Comment (felixfontein) [reply plugins/modules/source_control/gitlab/gitlab_project_variable.py]: Unfortunately GitHub doesn't want to show me what "this" is :)

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@markuman thanks a lot for implementing this!
@sguarin thanks for reviewing!
